### ISSUE TITLE
Doctrine Manager not initiliazed in phpunit with php 8.2

### ISSUE THREAD

**Issue by igoooor** [symfony/symfony#48666]
### Symfony version(s) affected

6.2.0

### Description

I am running php 8.2 with symfony 6.2.
I have a KernelTestCase which tries to load a repository:
```
self::bootKernel();
$container = static::getContainer();
$container->get(MyRepository::class);
```
and this results in the following error:
```
Error: Call to a member function isTransient() on null

/app/vendor/doctrine/persistence/src/Persistence/AbstractManagerRegistry.php:189
/app/vendor/doctrine/doctrine-bundle/Repository/ServiceEntityRepository.php:36
/app/src/Repository/MyRepository.php:36
/app/var/cache/publisher/test/ContainerJ3xN9um/Core_KernelTestDebugContainer.php:12920
/app/var/cache/publisher/test/ContainerJ3xN9um/Core_KernelTestDebugContainer.php:12056
/app/vendor/symfony/dependency-injection/Container.php:381
/app/vendor/symfony/dependency-injection/Argument/ServiceLocator.php:40
/app/vendor/symfony/doctrine-bridge/ContainerAwareEventManager.php:191
/app/vendor/symfony/doctrine-bridge/ContainerAwareEventManager.php:105
/app/vendor/doctrine/dbal/src/Connection.php:355
/app/vendor/doctrine/dbal/src/Connection.php:417
/app/vendor/doctrine/dbal/src/Connection.php:379
/app/vendor/doctrine/dbal/src/Connection.php:289
/app/vendor/doctrine/doctrine-bundle/ConnectionFactory.php:140
/app/vendor/doctrine/doctrine-bundle/ConnectionFactory.php:119
/app/var/cache/publisher/test/ContainerJ3xN9um/Core_KernelTestDebugContainer.php:1872
/app/var/cache/publisher/test/ContainerJ3xN9um/Core_KernelTestDebugContainer.php:1889
/app/vendor/symfony/var-exporter/Internal/LazyObjectState.php:95
/app/vendor/symfony/var-exporter/LazyGhostTrait.php:174
/app/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:862
/app/vendor/symfony/doctrine-bridge/CacheWarmer/ProxyCacheWarmer.php:50
/app/vendor/symfony/http-kernel/CacheWarmer/CacheWarmerAggregate.php:96
/app/vendor/symfony/http-kernel/Kernel.php:526
/app/vendor/symfony/http-kernel/Kernel.php:709
/app/vendor/symfony/http-kernel/Kernel.php:122
/app/vendor/symfony/framework-bundle/Test/KernelTestCase.php:73
/app/tests/app/publisher/Service/UserRoleRightServiceTest.php:20
```
`MyRepository.php:36` is just the constructor of the repository:
```
public function __construct(ManagerRegistry $registry)
{
    parent::__construct($registry, MyEntity::class);
}
```

If I `dump($manager);` in `vendor/doctrine/persistence/src/Persistence/AbstractManagerRegistry.php:187` I have the following output:
```
^ ContainerXmKqPiW\EntityManager_bad7396^ {#4157
  -config: null
  -conn: null
  -metadataFactory: null
  -unitOfWork: null
  -eventManager: null
  -proxyFactory: null
  -repositoryFactory: null
  -expressionBuilder: null
  -closed: false
  -filterCollection: null
  -cache: null
  -lazyObjectState: Symfony\Component\VarExporter\Internal\LazyObjectState^ {#15880
    +status: INITIALIZED_FULL
     …2
  }
}
```
This all worked previously with symfony 6.1 and php 8.1

### How to reproduce

Have an environment running with php 8.2 and symfony 6.2
Composer dependencies:
```
"phpunit/phpunit": "^9.0.0",
"symfony/phpunit-bridge": "6.2.*",
```

Have a unit test with:
```
self::bootKernel();
$container = static::getContainer();
$container->get(MyRepository::class);
```

run your test with `bin/phpunit`

**Comment by derrabus** [comment]
Please create a reproducer as described here: https://symfony.com/doc/current/contributing/code/reproducer.html

**Comment by igoooor** [comment]
yes I will, thanks for the guidance

**Comment by dunglas** [comment]
I'm having the same problem with a (complex) project I work on but having a hard time understanding what's going on. Are you using DAMA Doctrine Test Bundle on your project @igoooor?

**Comment by dunglas** [comment]
This doesn't look related to PHP 8.2 by the way, my project is still on 8.1.

**Comment by igoooor** [comment]
So far I have not been able to reproduce and a blank project, it only happens on my complex project since symfony 6.2.
What I noticed though, is that in my tests that were failing, I had the Kernel created by me (in a overwritten `createKernel()`, with `debug=true`. If I change to `debug=false` I don't have the issue anymore.
But with symfony 6.1, that did not make a difference, it worked with both values

**Comment by dunglas** [comment]
We investigated this issue privately with @nicolas-grekas. This looks related to the new proxy feature of Var Exporter. As a temporary workaround, switching back to Proxy Manager fix the issue. This can be done like that:

```php
// src/Kernel.php
// ...
    protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container, string $class, string $baseClass)
    {
        // cache the container
        $dumper = new PhpDumper($container);
        // The hack is here
        $dumper->setProxyDumper(new ProxyDumper());

        $content = $dumper->dump([
            'class' => $class,
            'base_class' => $baseClass,
            'file' => $cache->getPath(),
            'as_files' => true,
            'debug' => $this->debug,
            'build_time' => $container->hasParameter('kernel.container_build_time') ? $container->getParameter('kernel.container_build_time') : time(),
            'preload_classes' => array_map('get_class', $this->bundles),
        ]);

        $rootCode = array_pop($content);
        $dir = \dirname($cache->getPath()).'/';
        $fs = new Filesystem();

        foreach ($content as $file => $code) {
            $fs->dumpFile($dir.$file, $code);
            @chmod($dir.$file, 0666 & ~umask());
        }
        $legacyFile = \dirname($dir.key($content)).'.legacy';
        if (is_file($legacyFile)) {
            @unlink($legacyFile);
        }

        $cache->write($rootCode, $container->getResources());
    }
```

**Comment by nicolas-grekas** [comment]
I spent almost 2 days on this and I'm not done yet :sweat: 

This is triggered by a runtime circular loop:
- service A calls `$em->getClassMetadata` in its constructor,
- `$em` is lazy so this trigger a call to `em->__get(metadataFactory)`, which triggers initialization of `$em`
- nested inside this initializer is service B, who also calls `$em->getClassMetadata`
- but now since we're nested inside `$em->__get()` already, the magic method is not called anymore as the php engine guards against such recursive calls.
- :boom: 

This means that ghost objects are fundamentally incompatible with lazy objects involved in circular service graphs (aka #35345 is impossible in the generic case).

The main culprit for these runtime circular loops is ServiceEntityRepository.
I'm therefore proposing https://github.com/doctrine/DoctrineBundle/pull/1599 to make this class natively lazy on Symfony 6.2.

At a lower level, we could still ensure that such loops are resolved appropriately at runtime but this requires non trivial changes:

The first step is to force a virtual proxy for those lazy objects. And then, because the initialization of virtual proxies also relies on magic methods, they need to be reimplemented using regular initializers.

I've started a branch doing so but I'm putting this work on a halt until https://github.com/doctrine/DoctrineBundle/pull/1599 is deployed, waiting for more reports. The reason I'm following this approach is that 1. the task is non-trivial as I just wrote and 2. such runtime circular graphs are not legit to me. Yes, they can be made to work thanks to the DI component, but constructors should do less in the first place. So I'd better wait and discuss case by case, asking ppl to improve their constructors instead.
